src/devices/vdml_ext_adi.c

```diff
@@ -0,0 +1,416 @@
+/**
+ * \file devices/vdml_adi.c
+ *
+ * Contains functions for interacting with the V5 ADI.
+ *
+ * Copyright (c) 2017-2020, Purdue University ACM SIGBots.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <errno.h>
+#include <math.h>
+#include <stdio.h>
+
+#include "pros/ext_adi.h"
+#include "pros/adi.h"
+#include "kapi.h"
+#include "v5_api.h"
+#include "vdml/registry.h"
+#include "vdml/vdml.h"
+
+#define INTERNAL_ADI_PORT (22 - 1)
+
+#define ADI_MOTOR_MAX_SPEED 127
+#define ADI_MOTOR_MIN_SPEED -128
+
+#define NUM_MAX_TWOWIRE 4
+
+// Theoretical calibration time is 1024ms, but in practice this seemed to be the
+// actual time that it takes.
+#define GYRO_CALIBRATION_TIME 1300
+
+typedef union adi_data {
+	struct {
+		int32_t calib;
+	} analog_data;
+	struct {
+		bool was_pressed;
+	} digital_data;
+	struct {
+		bool reversed;
+	} encoder_data;
+	struct __attribute__((packed)) {
+		double multiplier;
+		double tare_value;
+	} gyro_data;
+} adi_data_s_t;
+
+#define transform_adi_port(port)       \
+	if (port >= 'a' && port <= 'h')      \
+		port -= 'a';                       \
+	else if (port >= 'A' && port <= 'H') \
+		port -= 'A';                       \
+	else                                 \
+		port--;                            \
+	if (port > 7 || port < 0) {          \
+		errno = ENXIO;                     \
+		return PROS_ERR;                   \
+	}
+
+#define validate_type(device, port, type)                                                                 \
+	adi_port_config_e_t config = (adi_port_config_e_t)vexDeviceAdiPortConfigGet(device->device_info, port); \
+	if (config != type) {                                                                                   \
+		errno = EADDRINUSE;                                                                                   \
+		return PROS_ERR;                                                                                      \
+	}
+
+#define validate_motor(device, port)                                                                      \
+	adi_port_config_e_t config = (adi_port_config_e_t)vexDeviceAdiPortConfigGet(device->device_info, port); \
+	if (config != E_ADI_LEGACY_PWM && config != E_ADI_LEGACY_SERVO) {                                       \
+		errno = EADDRINUSE;                                                                                   \
+		return PROS_ERR;                                                                                      \
+	}
```

Comment:
Alex and I talked about this a bit on slack/discord. I'd prefer to move away from these macros as they can cause a lot of error (I'll comment on atleast some of them I saw). You can commit this and I can fix it up later as I don't want to hold up the pr on this, or you can merge in some of the changes from the branch I shared earlier.

At the very least, they shouldn't be duplicated